android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt

```diff
@@ -2,55 +2,85 @@
 // SPDX-License-Identifier: BSD-3-Clause
 package com.tailscale.ipn
 
-import android.content.Context
 import android.net.ConnectivityManager
 import android.net.LinkProperties
 import android.net.Network
 import android.net.NetworkCapabilities
-import android.net.NetworkRequest
 import android.util.Log
 import libtailscale.Libtailscale
 import java.net.InetAddress
-import java.net.NetworkInterface
 
 object NetworkChangeCallback {
 
+  // Track the current active network
+  @Volatile private var activeNetwork: Network? = null
+  private val TAG = "NetworkChangeCallback"
+
+  // Create a lock object for synchronizing access to activeNetwork
+  private val activeNetworkLock = Any()
+
   // requestNetwork attempts to find the best network that matches the passed NetworkRequest. It is
   // possible that this might return an unusuable network, eg a captive portal.
   fun monitorDnsChanges(connectivityManager: ConnectivityManager, dns: DnsConfig) {
-    val networkConnectivityRequest =
-        NetworkRequest.Builder()
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
-            .build()
-
-    connectivityManager.registerNetworkCallback(
-        networkConnectivityRequest,
+
+    connectivityManager.registerDefaultNetworkCallback(
         object : ConnectivityManager.NetworkCallback() {
           override fun onAvailable(network: Network) {
-            super.onAvailable(network)
+            val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
```

Comment:
I don't think this is safe; from [the docs](https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onAvailable(android.net.Network)):

> Do NOT call ConnectivityManager.getNetworkCapabilities(android.net.Network) or ConnectivityManager.getLinkProperties(android.net.Network) or other synchronous ConnectivityManager methods in this callback as this is prone to race conditions (there is no guarantee the objects returned by these methods will be current). Instead, wait for a call to onCapabilitiesChanged(android.net.Network, android.net.NetworkCapabilities) and onLinkPropertiesChanged(android.net.Network, android.net.LinkProperties) whose arguments are guaranteed to be well-ordered with respect to other callbacks.

Reply:
thanks for calling this out!!!